File: platforms/ios/Sources/Feedback/IngestClient.swift
import Foundation

/// Distinguishes "retry later" from "never retry", exactly as the other platforms do.
public struct IngestError: Error {
    public let message: String
    public let status: Int?
    public let retryable: Bool
}

/// The queue's dependency on the network, narrowed to one method, so the outbox's
/// failure handling is testable with no simulator and no sockets.
public protocol SubmissionTransport: AnyObject {
    func submit(submissionJSON: Data, idempotencyKey: String) throws
}

/**
 HTTP transport built on URLSession. Blocking by design: the queue runs on its own
 serial queue, and the SDK never assumes it may use the host app's concurrency.
 */
public final class IngestClient: SubmissionTransport {
    private let endpoint: String
    private let projectId: String
    private let timeout: TimeInterval
    private let session: URLSession

    public init(endpoint: String, projectId: String, timeout: TimeInterval = 15) {
        // Trailing slashes would produce "//v1/reports", which some proxies reject.
        var trimmed = endpoint
        while trimmed.hasSuffix("/") { trimmed.removeLast() }
        self.endpoint = trimmed
        self.projectId = projectId
        self.timeout = timeout

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        // The SDK must not consume the user's cellular data on a background flush
        // any more than necessary; leave the system's discretion in place.
        config.waitsForConnectivity = false
        self.session = URLSession(configuration: config)
    }

    public func submit(submissionJSON: Data, idempotencyKey: String) throws {
        try post(
            path: "/v1/reports",
            body: submissionJSON,
            extraHeaders: ["idempotency-key": idempotencyKey]
        )
    }

    public func sendReplayChunk(_ chunkJSON: Data) throws {
        try post(path: "/v1/replay/chunks", body: chunkJSON, extraHeaders: [:])
    }

    private func post(path: String, body: Data, extraHeaders: [String: String]) throws {
        guard let url = URL(string: endpoint + path) else {
            throw IngestError(message: "Invalid endpoint", status: nil, retryable: false)
        }

        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "content-type")
        request.setValue(projectId, forHTTPHeaderField: "x-project-id")
        for (key, value) in extraHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }
        request.httpBody = body

        // Bridges URLSession's async API to this synchronous one. The caller is
        // always on the SDK's own serial queue, never the main thread.
        let semaphore = DispatchSemaphore(value: 0)
        var response: URLResponse?
        var transportError: Error?

        let task = session.dataTask(with: request) { _, res, err in
            response = res
            transportError = err
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()

        if let transportError {
            // No response reached us, so a retry is worth it.
            throw IngestError(message: transportError.localizedDescription, status: nil, retryable: true)
        }

        guard let http = response as? HTTPURLResponse else {
            throw IngestError(message: "No HTTP response", status: nil, retryable: true)
        }

        guard (200...299).contains(http.statusCode) else {
            // 408/429 are client-status but genuinely transient.
            let retryable = http.statusCode >= 500 || http.statusCode == 408 || http.statusCode == 429
            throw IngestError(
                message: "Ingest failed with \(http.statusCode)",
                status: http.statusCode,
                retryable: retryable
            )
        }
    }
}

File: platforms/android/feedback/src/main/kotlin/io/markerusa/feedback/IngestClient.kt
package io.markerusa.feedback

import java.io.BufferedOutputStream
import java.net.HttpURLConnection
import java.net.URL

/** Distinguishes "retry later" from "never retry", exactly as the web client does. */
class IngestException(
    message: String,
    val status: Int?,
    val retryable: Boolean
) : Exception(message)

/**
 * The queue's dependency on the network, narrowed to one method.
 *
 * An interface rather than the concrete client so the outbox's failure handling —
 * the logic most worth testing — is testable on the JVM with no device, no server,
 * and no sockets.
 */
interface SubmissionTransport {
    fun submit(submissionJson: String, idempotencyKey: String)
}

/**
 * HTTP transport. Uses HttpURLConnection to keep the library dependency-free —
 * see the note in build.gradle.kts about SDKs and version conflicts.
 *
 * Every method here blocks. Callers are responsible for threading; the SDK never
 * assumes a coroutine scope or an executor belonging to the host app.
 */
class IngestClient(
    endpoint: String,
    private val projectId: String,
    private val timeoutMs: Int = 15_000
) : SubmissionTransport {
    private val endpoint = endpoint.trimEnd('/')

    override fun submit(submissionJson: String, idempotencyKey: String) {
        post("/v1/reports", submissionJson, mapOf("idempotency-key" to idempotencyKey))
    }

    fun sendReplayChunk(chunkJson: String) {
        post("/v1/replay/chunks", chunkJson, emptyMap())
    }

    private fun post(path: String, body: String, extraHeaders: Map<String, String>) {
        var connection: HttpURLConnection? = null
        try {
            connection = (URL("$endpoint$path").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = timeoutMs
                readTimeout = timeoutMs
                doOutput = true
                setRequestProperty("content-type", "application/json")
                setRequestProperty("x-project-id", projectId)
                extraHeaders.forEach { (k, v) -> setRequestProperty(k, v) }
            }

            BufferedOutputStream(connection.outputStream).use { out ->
                out.write(body.toByteArray(Charsets.UTF_8))
                out.flush()
            }

            val status = connection.responseCode
            if (status !in 200..299) {
                // 408/429 are client-status but genuinely transient.
                val retryable = status >= 500 || status == 408 || status == 429
                // Drain the error stream so the connection can be pooled rather
                // than torn down — otherwise every failure leaks a socket.
                connection.errorStream?.use { it.readBytes() }
                throw IngestException("Ingest failed with $status", status, retryable)
            }

            connection.inputStream.use { it.readBytes() }
        } catch (e: IngestException) {
            throw e
        } catch (e: Exception) {
            // No response reached us, so a retry is worth it.
            throw IngestException(e.message ?: "Network error", null, true)
        } finally {
            connection?.disconnect()
        }
    }
}
